libs/sr-opt/src/parentWorker.ts

```diff
@@ -0,0 +1,177 @@
+import type { ChildCommand, ChildMessage } from './childWorker'
+import type { BuildResult } from './optimize'
+
+export interface ParentCommandStart {
+  command: 'start'
+  relicsBySlot: Record<RelicSlotKey, ICachedRelic[]>
+  detachedNodes: NumTagFree[]
+  numWorkers: number
+  maxResults: number
+}
+export type ParentCommand = ParentCommandStart
+
+export interface ParentMessageProgress {
+  resultType: 'progress'
+  numBuilds: number
+  numBuildsComputed: number
+  sendingResults?: boolean
+}
+export interface ParentMessageSending {
+  resultType: 'sending'
+}
+export interface ParentMessageDone {
+  resultType: 'done'
+  buildResults: BuildResult[]
+}
+export interface ParentMessageErr {
+  resultType: 'err'
+  message: string
+}
+export type ParentMessage =
+  | ParentMessageProgress
+  | ParentMessageDone
+  | ParentMessageErr
+
+export type RelicStats = {
+  id: string
+  stats: Record<string, number>
+}
+
+// Get proper typings for posting a message back to main thread
+declare function postMessage(message: ParentMessage): void
+
+// Receiving a message from main thread to worker
+onmessage = async (e: MessageEvent<ParentCommand>) => {
+  try {
+    await handleEvent(e)
+  } catch (err) {
+    console.log(err)
+    postMessage({ resultType: 'err', message: JSON.stringify(err) })
+  }
+}
+
+async function handleEvent(e: MessageEvent<ParentCommand>): Promise<void> {
+  const { data } = e,
+    { command } = data
+  switch (command) {
+    case 'start':
+      await start(data)
+  }
+}
+
+async function start({
+  relicsBySlot,
+  detachedNodes,
+  numWorkers,
+  maxResults,
+}: ParentCommandStart) {
+  // Step 3: Optimize nodes, as needed
+  const flatNodes = flatten(detachedNodes)
+  const combinedNodes = combineConst(flatNodes)
+
+  const relicStatsBySlot = objKeyMap(allRelicSlotKeys, (slot) =>
+    relicsBySlot[slot].map(convertRelicToStats)
+  )
+
+  const chunkSize = Math.ceil(relicsBySlot.head.length / numWorkers)
+  // Spawn child workers to calculate builds
+  const workers = range(1, numWorkers).map(
+    () =>
+      new Worker(new URL('./childWorker.ts', import.meta.url), {
+        type: 'module',
+      })
+  )
+  // Wait for all workers to finish optimizing
+  let results: BuildResult[] = []
+  let numBuildsComputed = 0
+  await Promise.all(
+    workers.map((worker, index) => {
+      return new Promise<void>((res, rej) => {
+        // On worker completion, resolve promise
+        worker.onmessage = ({ data }: MessageEvent<ChildMessage>) => {
+          switch (data.resultType) {
+            case 'initialized':
+              // Worker is initialized; start optimizing
+              worker.postMessage({ command: 'start' })
+              break
+            case 'results':
+              numBuildsComputed += data.builds.length
+              results = results.concat(data.builds)
+              // TODO: This is slow, might need to think of something else
+              if (results.length > maxResults) {
```

Comment:
Should we keep a sorted array and have customer inserter/merger? Not sure if it would be any faster

Reply:
I think this is fine, and looks like this happens in worker space, and isnt a hot zone. I'd say keep this simple is better than writing something hard to debug/maintain.

Reply:
We can set threshold to a value much larger than `maxResult`, and close to memory limit (and make sure it's still larger than `maxResult`). Otherwise, this will trigger on every loop after the first time, which is very slow.

Reply:
Also, if we keep track of the largest build after each sort, and use that to filter suboptimal builds, that should prevent `results` from growing too fast if good builds are found early.